File: Vista/DevolucionProveedorDialog.py
from PyQt6.QtWidgets import QDialog, QVBoxLayout, QHBoxLayout, QLabel, QLineEdit, QPushButton, QMessageBox, QWidget
from PyQt6.QtCore import Qt
from Vista.Estilos import Estilos

class DevolucionProveedorDialog(QDialog):
    def __init__(self, parent, p, ctrl):
        super().__init__(parent)
        self.ctrl = ctrl
        self.producto = p
        self.setWindowTitle("Devolucion a Proveedor")
        self.setModal(True)
        self.construir()

    def construir(self):
        self.setFixedSize(380, 300)
        panel = QWidget(self)
        panel.resize(380, 300)
        panel.setStyleSheet(f"background-color: {Estilos.BG_BLANCO.name()};")

        tit = QLabel("DEVOLVER PRODUCTOS", panel)
        tit.setFont(Estilos.FUENTE_TITULO)
        tit.setGeometry(24, 18, 300, 30)

        lblProd = QLabel(f"Producto: {self.producto.getNombre()}", panel)
        lblProd.setFont(Estilos.FUENTE_BOLD)
        lblProd.setGeometry(24, 60, 330, 20)

        lblStock = QLabel(f"Stock actual: {self.producto.getStock()} {self.producto.getUnidad()}", panel)
        lblStock.setFont(Estilos.FUENTE_NORMAL)
        lblStock.setStyleSheet(f"color: {Estilos.TEXTO_SECUNDARIO.name()};")
        lblStock.setGeometry(24, 85, 330, 20)

        lblCant = QLabel("CANTIDAD A DEVOLVER:", panel)
        lblCant.setFont(Estilos.FUENTE_XS)
        lblCant.setStyleSheet(f"color: {Estilos.TEXTO_TENUE.name()}; border: none;")
        lblCant.setGeometry(24, 130, 300, 14)

        self.txtCantidad = QLineEdit(panel)
        self.txtCantidad.setFont(Estilos.FUENTE_BOLD)
        self.txtCantidad.setStyleSheet(f"""
            QLineEdit {{
                background-color: {Estilos.BG_ZINC_100.name()};
                border: 2px solid {Estilos.BORDE.name()};
                padding: 6px 10px;
            }}
        """)
        self.txtCantidad.setGeometry(24, 148, 315, 36)

        btnCancelar = Estilos.botonSecundario("Cancelar")
        btnCancelar.setParent(panel)
        btnCancelar.setGeometry(110, 210, 110, 38)
        btnCancelar.clicked.connect(self.reject)

        btnGuardar = Estilos.botonPeligro("Devolver")
        btnGuardar.setParent(panel)
        btnGuardar.setGeometry(230, 210, 110, 38)
        btnGuardar.clicked.connect(self.guardar)

    def guardar(self):
        try:
            cantidad = float(self.txtCantidad.text().strip())
            if cantidad <= 0:
                QMessageBox.warning(self, "Aviso", "Ingrese una cantidad valida mayor a 0")
                return
            if cantidad > self.producto.getStock():
                QMessageBox.warning(self, "Aviso", "No puede devolver mas del stock existente")
                return
                
            self.producto.setStock(self.producto.getStock() - cantidad)
            self.ctrl.guardarProducto(self.producto)
            QMessageBox.information(self, "Exito", "Devolucion procesada exitosamente")
            self.accept()
        except ValueError:
            QMessageBox.critical(self, "Error", "Numero invalido")

File: Vista/DevolucionProveedorDialog.cs
using Microsoft.Maui.Controls;
using Microsoft.Maui.Graphics;
using Microsoft.Maui.Layouts;
using System;
using System.Threading.Tasks;
using punto_de_venta_C_.Modelo;
using punto_de_venta_C_.Controlador;

namespace punto_de_venta_C_.Vista
{
    public class DevolucionProveedorDialog : ContentPage
    {
        private readonly InventarioControlador ctrl;
        private readonly Producto producto;
        private Entry txtCantidad;
        private TaskCompletionSource<bool> tcs = new TaskCompletionSource<bool>();

        public DevolucionProveedorDialog(Producto p, InventarioControlador ctrl)
        {
            this.ctrl = ctrl;
            this.producto = p;
            Title = "Devolucion a Proveedor";
            BackgroundColor = Estilos.BG_BLANCO;
            construir();
        }

        private void construir()
        {
            var absoluteLayout = new AbsoluteLayout
            {
                WidthRequest = 380,
                HeightRequest = 300,
                BackgroundColor = Estilos.BG_BLANCO
            };

            var tit = new Label
            {
                Text = "DEVOLVER PRODUCTOS",
                FontSize = Estilos.FUENTE_TITULO_SIZE,
                FontAttributes = FontAttributes.Bold,
                TextColor = Estilos.TEXTO_PRINCIPAL
            };
            AbsoluteLayout.SetLayoutBounds(tit, new Rect(24, 18, 300, 30));
            AbsoluteLayout.SetLayoutFlags(tit, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(tit);

            var lblProd = new Label
            {
                Text = "Producto: " + producto.getNombre(),
                FontSize = Estilos.FUENTE_BOLD_SIZE,
                FontAttributes = FontAttributes.Bold,
                TextColor = Estilos.TEXTO_PRINCIPAL
            };
            AbsoluteLayout.SetLayoutBounds(lblProd, new Rect(24, 60, 330, 20));
            AbsoluteLayout.SetLayoutFlags(lblProd, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(lblProd);

            var lblStock = new Label
            {
                Text = "Stock actual: " + producto.getStock() + " " + producto.getUnidad(),
                FontSize = Estilos.FUENTE_NORMAL_SIZE,
                TextColor = Estilos.TEXTO_SECUNDARIO
            };
            AbsoluteLayout.SetLayoutBounds(lblStock, new Rect(24, 85, 330, 20));
            AbsoluteLayout.SetLayoutFlags(lblStock, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(lblStock);

            var lblCant = new Label
            {
                Text = "CANTIDAD A DEVOLVER:",
                FontSize = Estilos.FUENTE_XS_SIZE,
                TextColor = Estilos.TEXTO_TENUE
            };
            AbsoluteLayout.SetLayoutBounds(lblCant, new Rect(24, 130, 300, 14));
            AbsoluteLayout.SetLayoutFlags(lblCant, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(lblCant);

            txtCantidad = new Entry
            {
                FontSize = Estilos.FUENTE_BOLD_SIZE,
                BackgroundColor = Estilos.BG_ZINC_100,
                TextColor = Estilos.TEXTO_PRINCIPAL,
                Keyboard = Keyboard.Numeric
            };
            AbsoluteLayout.SetLayoutBounds(txtCantidad, new Rect(24, 148, 315, 36));
            AbsoluteLayout.SetLayoutFlags(txtCantidad, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(txtCantidad);

            var btnCancelar = Estilos.botonSecundario("Cancelar");
            btnCancelar.Clicked += async (s, e) =>
            {
                tcs.SetResult(false);
                await Navigation.PopModalAsync();
            };
            AbsoluteLayout.SetLayoutBounds(btnCancelar, new Rect(110, 210, 110, 38));
            AbsoluteLayout.SetLayoutFlags(btnCancelar, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(btnCancelar);

            var btnGuardar = Estilos.botonPeligro("Devolver");
            btnGuardar.Clicked += (s, e) => intentarDevolucion();
            AbsoluteLayout.SetLayoutBounds(btnGuardar, new Rect(230, 210, 110, 38));
            AbsoluteLayout.SetLayoutFlags(btnGuardar, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(btnGuardar);

            Content = new Grid
            {
                Children = { absoluteLayout },
                HorizontalOptions = LayoutOptions.Center,
                VerticalOptions = LayoutOptions.Center
            };
        }

        private async void intentarDevolucion()
        {
            try
            {
                double cantidad = double.Parse(txtCantidad.Text?.Trim() ?? "0");
                if (cantidad <= 0)
                {
                    await DisplayAlert("Aviso", "Ingrese una cantidad valida mayor a 0", "OK");
                    return;
                }
                if (cantidad > producto.getStock())
                {
                    await DisplayAlert("Aviso", "No puede devolver mas del stock existente", "OK");
                    return;
                }
                producto.setStock(producto.getStock() - cantidad);
                ctrl.guardarProducto(producto);
                await DisplayAlert("Éxito", "Devolucion procesada exitosamente", "OK");
                tcs.SetResult(true);
                await Navigation.PopModalAsync();
            }
            catch (Exception)
            {
                await DisplayAlert("Error", "Numero invalido", "OK");
            }
        }

        public async Task<bool> mostrar()
        {
            if (Application.Current?.MainPage != null)
            {
                await Application.Current.MainPage.Navigation.PushModalAsync(new NavigationPage(this));
            }
            return await tcs.Task;
        }
    }
}
